storage/fanout.go

```diff
@@ -471,26 +453,45 @@ func newGenericMergeSeriesSet(sets []genericSeriesSet, querier *mergeGenericQuer
 		return sets[0]
 	}
 
-	// Sets need to be pre-advanced, so we can introspect the label of the
-	// series under the cursor.
-	var h genericSeriesSetHeap
-	for _, set := range sets {
-		if set == nil {
-			continue
-		}
-		if set.Next() {
-			heap.Push(&h, set)
-		}
-	}
 	return &genericMergeSeriesSet{
 		mergeFunc: mergeFunc,
-		heap:      h,
 		sets:      sets,
 		querier:   querier,
 	}
 }
 
 func (c *genericMergeSeriesSet) Next() bool {
+	if c.heap == nil {
+		// Sets need to be pre-advanced, so we can introspect the label of the
```

Comment:
This is only advancing this select, we need to advance all the selects done against the querier to spot if any of them failed so that we can filter out responses from remotes that fail some of their requests.

Reply:
Could you please elaborate this one a bit more?

Reply:
I think what Brian is saying you exactly did below @kakkoyun , so we make sure `Next` is done so we ensure no error is there.

However TBH we  can simplify this whole thing by one simple trick: We can make sure remote read Select will behave like this (if select fails we only return warning not error). This will make the flow super easy (: We can name it even `nonErrReadQuerier` or something.

Reply:
You're still only advancing the sets of this Select, not the sets of all Selects.